e2e/README.md

```diff
@@ -145,7 +149,1089 @@ func E2ETests(env e2e.TestEnv) testhelper.Tests {
+suite.Run() function (in e2e/internal/testhelper/testhelper.go), called by the
+main e2e.Run() function (in e2e/suite.go) does indeed call.
+
+You should therefore assume, when writing an individual e2e test, that it *will*
+run in parallel to other e2e tests.
+
+In general, you should try your best to make your test safe to run in parallel.
+For example, if your test involves building a new image file, don't just put the
+file in the current directory; create a [dedicated temporary
+directory](#temporary-dirs--files-and-cleanup) for this particular test under
+`testenv.TestDir`, and build & use your image by specifying an absolute path to
+your image file in that temporary subdir.
+
+With that said, it is still the case that *some* tests cannot be run in parallel
+to one another. Some examples include:
+
+- Tests that require changing the current working directory.
+  - Note that this does not include calls to
+    [e2e.WithDir()](#functional-options-to-testenvrunsingularity) in
+    testenv.RunSingularity(). These *are* safe to run in parallel, as they only
+    affect the singularity process that the test launches, not the process
+    running the test code itself.
+- Tests that require changing the OS
+  [umask](https://en.wikipedia.org/wiki/Umask).
+- Tests that affect files in the user's homedir (or in root's homedir, i.e.
+  `/root`).
+  - Even though the e2e suite sets up ["fake"
+    homedirs](#initialization-and-the-e2etestenv-struct) for the current user
+    and for root, those homedirs are still one and the same for the entire e2e
+    run. And so, if two different tests were to manipulate files in the homedir
+    at once, they could interfere with each other.
+  - Examples where this concern arises include any test that would potentially
+    change, or be sensitive to, the contents of files inside the user's
+    `$HOME/.singularity` directory, such as `remote.yaml`, `docker-config.json`,
+    and others, as well as any test that changes the content of the system
+    `singularity.conf`.
+
+To deal with such cases, e2e/internal/testhelper/testhelper.go defines a
+function `testhelper.NoParallel(func(*testing.T)) func(*testing.T)`. This
+function returns the argument it was given, but first adds it to a data
+structure that holds all the test functions that *cannot* be run in parallel.
+The suite.Run() function in e2e/internal/testhelper/testhelper.go then makes
+sure that these tests are run sequentially, and not in parallel with any other
+tests.
+
+It is for this reason that a typical test name in the e2e suite looks as follows:
+
+```plain
+<...>
+TestE2E/PAR/BUILD/build_with_bind_mount
+<...>
+TestE2E/SEQ/DOCKER/cred_prio
+<...>
+```
+
+`TestE2E` is the test name for the entire e2e suite; it is followed by `PAR`,
+for the set of tests run in parallel, or by `SEQ`, for those tests that cannot
+be run in parallel and are run sequentially.
+
+The fact that the testhelper.NoParallel() function returns its argument as its
```

Comment:
I've simplified the first half of the explanation (as per your previous comment). But I think the part about the return value is important, otherwise the way `testhelper.Tests` maps get written doesn't really make sense. If `Fn` is a function that "marks its argument as X", you would usually conclude from that that the way to use `Fn` is to call `Fn(arg)` (with no return value or an error return value), and then use `arg` when you need its value. `testhelper.NoParallel` is written in a way that this will also work, but I think that in a README, we want to encourage people to use the pattern of marking a non-parallel test "in place" —

```go
		"test help":      c.remoteTestHelp,
		"use":            c.remoteUse,
		"use exclusive":  np(c.remoteUseExclusive),
```